File: core/state_manager/state_manager.py
"""
状态管理模块

提供统一的状态管理功能，包括状态定义、状态转换、状态锁定等。

作者: AI Assistant
创建时间: 2024-10-24
版本: 1.0.0
"""

from typing import Dict, List, Optional
from enum import Enum
import time
import logging

logger = logging.getLogger(__name__)


class State(Enum):
    """状态枚举"""
    IDLE = 'idle'                    # 空闲状态
    TEXT_SSE = 'text_sse'           # 文本SSE处理中
    TEXT_TTS = 'text_tts'           # 文本TTS处理中
    VOICE_STT = 'voice_stt'         # 语音STT处理中
    VOICE_SSE = 'voice_sse'         # 语音SSE处理中
    VOICE_TTS = 'voice_tts'         # 语音TTS处理中
    VOICE_CALL = 'voice_call'       # 语音通话中
    ERROR = 'error'                  # 错误状态


class StateManager:
    """统一状态管理器"""
    
    def __init__(self):
        """初始化状态管理器"""
        self.current_state = State.IDLE
        self.state_history = []
        self.state_locked = False
        self.state_transitions = {
            State.IDLE: [State.TEXT_SSE, State.VOICE_STT, State.VOICE_CALL],
            State.TEXT_SSE: [State.TEXT_TTS, State.IDLE, State.ERROR],
            State.TEXT_TTS: [State.IDLE, State.ERROR],
            State.VOICE_STT: [State.VOICE_SSE, State.IDLE, State.ERROR],
            State.VOICE_SSE: [State.VOICE_TTS, State.IDLE, State.ERROR],
            State.VOICE_TTS: [State.IDLE, State.ERROR],
            State.VOICE_CALL: [State.IDLE, State.ERROR],
            State.ERROR: [State.IDLE]
        }
        self.state_change_callbacks = []
        self.state_lock_time = None
        self.max_lock_duration = 300  # 5分钟最大锁定时间
    
    def set_state(self, new_state: State) -> bool:
        """
        设置状态
        
        Args:
            new_state: 新状态
        
        Returns:
            是否设置成功
        """
        if self.state_locked:
            logger.warning(f"状态已锁定，无法从 {self.current_state.value} 转换到 {new_state.value}")
            return False
        
        if not self.can_transition(new_state):
            logger.warning(f"无效的状态转换：从 {self.current_state.value} 到 {new_state.value}")
            return False
        
        # 记录状态历史
        previous_state = self.current_state
        self.state_history.append({
            'from': previous_state,
            'to': new_state,
            'timestamp': time.time()
        })
        
        # 更新状态
        self.current_state = new_state
        
        # 触发状态变化回调
        for callback in self.state_change_callbacks:
            try:
                callback(previous_state, new_state)
            except Exception as e:
                logger.error(f"状态变化回调执行失败: {e}")
        
        logger.info(f"状态转换成功: {previous_state.value} -> {new_state.value}")
        return True
    
    def get_state(self) -> State:
        """
        获取当前状态
        
        Returns:
            当前状态
        """
        return self.current_state
    
    def can_transition(self, new_state: State) -> bool:
        """
        检查状态转换是否合法
        
        Args:
            new_state: 目标状态
        
        Returns:
            是否可以转换
        """
        return new_state in self.state_transitions.get(self.current_state, [])
    
    def lock_state(self, duration: Optional[int] = None) -> None:
        """
        锁定状态
        
        Args:
            duration: 锁定持续时间（秒），None表示永久锁定
        """
        self.state_locked = True
        self.state_lock_time = time.time()
        if duration:
            self.max_lock_duration = duration
        logger.info(f"状态已锁定，持续时间: {duration}秒" if duration else "状态已永久锁定")
    
    def unlock_state(self) -> None:
        """解锁状态"""
        self.state_locked = False
        self.state_lock_time = None
        logger.info("状态已解锁")
    
    def is_state_locked(self) -> bool:
        """
        检查状态是否被锁定
        
        Returns:
            是否被锁定
        """
        if not self.state_locked:
            return False
        
        # 检查锁定是否超时
        if self.state_lock_time and time.time() - self.state_lock_time > self.max_lock_duration:
            logger.warning("状态锁定超时，自动解锁")
            self.unlock_state()
            return False
        
        return True
    
    def rollback_state(self) -> bool:
        """
        回滚状态
        
        Returns:
            是否回滚成功
        """
        if not self.state_history:
            logger.warning("没有状态历史可以回滚")
            return False
        
        previous_state_info = self.state_history.pop()
        previous_state = previous_state_info['from']
        
        # 直接设置状态，不检查转换规则
        self.current_state = previous_state
        
        logger.info(f"状态已回滚到: {previous_state.value}")
        return True
    
    def get_state_history(self) -> List[Dict]:
        """
        获取状态历史
        
        Returns:
            状态历史列表
        """
        return self.state_history.copy()
    
    def clear_state_history(self) -> None:
        """清空状态历史"""
        self.state_history.clear()
        logger.info("状态历史已清空")
    
    def register_state_change_callback(self, callback) -> None:
        """
        注册状态变化回调
        
        Args:
            callback: 回调函数，接收 (from_state, to_state) 参数
        """
        self.state_change_callbacks.append(callback)
        logger.info("状态变化回调已注册")
    
    def unregister_state_change_callback(self, callback) -> None:
        """
        注销状态变化回调
        
        Args:
            callback: 要注销的回调函数
        """
        if callback in self.state_change_callbacks:
            self.state_change_callbacks.remove(callback)
            logger.info("状态变化回调已注销")
    
    def get_available_transitions(self) -> List[State]:
        """
        获取可用的状态转换
        
        Returns:
            可转换的状态列表
        """
        return self.state_transitions.get(self.current_state, [])
    
    def force_set_state(self, new_state: State) -> bool:
        """
        强制设置状态（忽略转换规则和锁定）
        
        Args:
            new_state: 新状态
        
        Returns:
            是否设置成功
        """
        previous_state = self.current_state
        self.current_state = new_state
        
        # 记录强制状态变化
        self.state_history.append({
            'from': previous_state,
            'to': new_state,
            'timestamp': time.time(),
            'forced': True
        })
        
        logger.warning(f"强制状态转换: {previous_state.value} -> {new_state.value}")
        return True
    
    def reset_to_idle(self) -> bool:
        """
        重置到空闲状态
        
        Returns:
            是否重置成功
        """
        if self.current_state == State.IDLE:
            return True
        
        # 清空历史并强制设置为IDLE
        self.state_history.clear()
        self.state_locked = False
        self.current_state = State.IDLE
        
        logger.info("状态已重置到空闲状态")
        return True
    
    def get_state_info(self) -> Dict:
        """
        获取状态信息
        
        Returns:
            状态信息字典
        """
        return {
            'current_state': self.current_state.value,
            'is_locked': self.is_state_locked(),
            'lock_time': self.state_lock_time,
            'available_transitions': [state.value for state in self.get_available_transitions()],
            'history_count': len(self.state_history),
            'max_lock_duration': self.max_lock_duration
        }


# 便捷函数
def create_state_manager() -> StateManager:
    """
    创建状态管理器
    
    Returns:
        状态管理器实例
    """
    return StateManager()


def get_state_name(state: State) -> str:
    """
    获取状态名称
    
    Args:
        state: 状态枚举
    
    Returns:
        状态名称
    """
    return state.value


def is_valid_state(state_name: str) -> bool:
    """
    检查状态名称是否有效
    
    Args:
        state_name: 状态名称
    
    Returns:
        是否有效
    """
    try:
        State(state_name)
        return True
    except ValueError:
        return False

File: cleanup/backup_files/js/state_manager.js
/**
 * JavaScript状态管理器
 * 
 * 提供与Python状态管理器对应的JavaScript实现，包括状态定义、状态转换、状态锁定等。
 * 
 * 作者: AI Assistant
 * 创建时间: 2024-10-24
 * 版本: 1.0.0
 */

class StateManager {
    /**
     * 构造函数
     */
    constructor() {
        this.currentState = 'idle';
        this.stateHistory = [];
        this.stateLocked = false;
        this.stateTransitions = {
            'idle': ['text_sse', 'voice_stt', 'voice_call'],
            'text_sse': ['text_tts', 'idle', 'error'],
            'text_tts': ['idle', 'error'],
            'voice_stt': ['voice_sse', 'idle', 'error'],
            'voice_sse': ['voice_tts', 'idle', 'error'],
            'voice_tts': ['idle', 'error'],
            'voice_call': ['idle', 'error'],
            'error': ['idle']
        };
        this.stateChangeCallbacks = [];
        this.stateLockTime = null;
        this.maxLockDuration = 300000; // 5分钟最大锁定时间（毫秒）
    }
    
    /**
     * 设置状态
     * @param {string} newState - 新状态
     * @returns {boolean} 是否设置成功
     */
    setState(newState) {
        if (this.stateLocked) {
            console.warn(`状态已锁定，无法从 ${this.currentState} 转换到 ${newState}`);
            return false;
        }
        
        if (!this.canTransition(newState)) {
            console.warn(`无效的状态转换：从 ${this.currentState} 到 ${newState}`);
            return false;
        }
        
        // 记录状态历史
        const previousState = this.currentState;
        this.stateHistory.push({
            from: previousState,
            to: newState,
            timestamp: Date.now()
        });
        
        // 更新状态
        this.currentState = newState;
        
        // 触发状态变化回调
        this.stateChangeCallbacks.forEach(callback => {
            try {
                callback(previousState, newState);
            } catch (error) {
                console.error('状态变化回调执行失败:', error);
            }
        });
        
        console.log(`状态转换成功: ${previousState} -> ${newState}`);
        return true;
    }
    
    /**
     * 获取当前状态
     * @returns {string} 当前状态
     */
    getState() {
        return this.currentState;
    }
    
    /**
     * 检查状态转换是否合法
     * @param {string} newState - 目标状态
     * @returns {boolean} 是否可以转换
     */
    canTransition(newState) {
        return this.stateTransitions[this.currentState]?.includes(newState) || false;
    }
    
    /**
     * 锁定状态
     * @param {number} duration - 锁定持续时间（毫秒），undefined表示永久锁定
     */
    lockState(duration) {
        this.stateLocked = true;
        this.stateLockTime = Date.now();
        if (duration) {
            this.maxLockDuration = duration;
        }
        console.log(`状态已锁定，持续时间: ${duration ? duration + 'ms' : '永久'}`);
    }
    
    /**
     * 解锁状态
     */
    unlockState() {
        this.stateLocked = false;
        this.stateLockTime = null;
        console.log('状态已解锁');
    }
    
    /**
     * 检查状态是否被锁定
     * @returns {boolean} 是否被锁定
     */
    isStateLocked() {
        if (!this.stateLocked) {
            return false;
        }
        
        // 检查锁定是否超时
        if (this.stateLockTime && Date.now() - this.stateLockTime > this.maxLockDuration) {
            console.warn('状态锁定超时，自动解锁');
            this.unlockState();
            return false;
        }
        
        return true;
    }
    
    /**
     * 回滚状态
     * @returns {boolean} 是否回滚成功
     */
    rollbackState() {
        if (this.stateHistory.length === 0) {
            console.warn('没有状态历史可以回滚');
            return false;
        }
        
        const previousStateInfo = this.stateHistory.pop();
        const previousState = previousStateInfo.from;
        
        // 直接设置状态，不检查转换规则
        this.currentState = previousState;
        
        console.log(`状态已回滚到: ${previousState}`);
        return true;
    }
    
    /**
     * 获取状态历史
     * @returns {Array} 状态历史列表
     */
    getStateHistory() {
        return [...this.stateHistory];
    }
    
    /**
     * 清空状态历史
     */
    clearStateHistory() {
        this.stateHistory = [];
        console.log('状态历史已清空');
    }
    
    /**
     * 注册状态变化回调
     * @param {Function} callback - 回调函数，接收 (fromState, toState) 参数
     */
    registerStateChangeCallback(callback) {
        this.stateChangeCallbacks.push(callback);
        console.log('状态变化回调已注册');
    }
    
    /**
     * 注销状态变化回调
     * @param {Function} callback - 要注销的回调函数
     */
    unregisterStateChangeCallback(callback) {
        const index = this.stateChangeCallbacks.indexOf(callback);
        if (index > -1) {
            this.stateChangeCallbacks.splice(index, 1);
            console.log('状态变化回调已注销');
        }
    }
    
    /**
     * 获取可用的状态转换
     * @returns {Array} 可转换的状态列表
     */
    getAvailableTransitions() {
        return this.stateTransitions[this.currentState] || [];
    }
    
    /**
     * 强制设置状态（忽略转换规则和锁定）
     * @param {string} newState - 新状态
     * @returns {boolean} 是否设置成功
     */
    forceSetState(newState) {
        const previousState = this.currentState;
        this.currentState = newState;
        
        // 记录强制状态变化
        this.stateHistory.push({
            from: previousState,
            to: newState,
            timestamp: Date.now(),
            forced: true
        });
        
        console.warn(`强制状态转换: ${previousState} -> ${newState}`);
        return true;
    }
    
    /**
     * 重置到空闲状态
     * @returns {boolean} 是否重置成功
     */
    resetToIdle() {
        if (this.currentState === 'idle') {
            return true;
        }
        
        // 清空历史并强制设置为IDLE
        this.stateHistory = [];
        this.stateLocked = false;
        this.currentState = 'idle';
        
        console.log('状态已重置到空闲状态');
        return true;
    }
    
    /**
     * 获取状态信息
     * @returns {Object} 状态信息字典
     */
    getStateInfo() {
        return {
            currentState: this.currentState,
            isLocked: this.isStateLocked(),
            lockTime: this.stateLockTime,
            availableTransitions: this.getAvailableTransitions(),
            historyCount: this.stateHistory.length,
            maxLockDuration: this.maxLockDuration
        };
    }
    
    /**
     * 检查状态是否为特定值
     * @param {string} state - 状态值
     * @returns {boolean} 是否匹配
     */
    isState(state) {
        return this.currentState === state;
    }
    
    /**
     * 检查状态是否在特定列表中
     * @param {Array} states - 状态列表
     * @returns {boolean} 是否在列表中
     */
    isStateIn(states) {
        return states.includes(this.currentState);
    }
    
    /**
     * 获取状态持续时间
     * @returns {number} 持续时间（毫秒）
     */
    getStateDuration() {
        if (this.stateHistory.length === 0) {
            return 0;
        }
        
        const lastTransition = this.stateHistory[this.stateHistory.length - 1];
        return Date.now() - lastTransition.timestamp;
    }
    
    /**
     * 获取状态统计
     * @returns {Object} 状态统计
     */
    getStateStats() {
        const stats = {};
        this.stateHistory.forEach(transition => {
            stats[transition.to] = (stats[transition.to] || 0) + 1;
        });
        return stats;
    }
}

// 便捷函数
/**
 * 创建状态管理器
 * @returns {StateManager} 状态管理器实例
 */
function createStateManager() {
    return new StateManager();
}

/**
 * 获取状态名称
 * @param {string} state - 状态值
 * @returns {string} 状态名称
 */
function getStateName(state) {
    return state;
}

/**
 * 检查状态名称是否有效
 * @param {string} stateName - 状态名称
 * @returns {boolean} 是否有效
 */
function isValidState(stateName) {
    const validStates = [
        'idle', 'text_sse', 'text_tts', 'voice_stt', 
        'voice_sse', 'voice_tts', 'voice_call', 'error'
    ];
    return validStates.includes(stateName);
}

// 全局状态管理器实例
window.stateManager = new StateManager();

// 便捷函数
/**
 * 获取配置值的便捷函数
 * @param {string} key - 配置键
 * @param {*} defaultValue - 默认值
 * @returns {*} 配置值
 */
window.getState = function() {
    return window.stateManager.getState();
};

/**
 * 设置状态的便捷函数
 * @param {string} state - 状态值
 * @returns {boolean} 是否设置成功
 */
window.setState = function(state) {
    return window.stateManager.setState(state);
};

/**
 * 检查状态的便捷函数
 * @param {string} state - 状态值
 * @returns {boolean} 是否匹配
 */
window.isState = function(state) {
    return window.stateManager.isState(state);
};

// 页面加载完成后初始化
document.addEventListener('DOMContentLoaded', function() {
    console.log('JavaScript状态管理器已初始化');
    console.log('当前状态:', window.stateManager.getState());
});
